src/votekit/ballot_generator.py

```diff
@@ -0,0 +1,534 @@
+
+                    # print(ballot)
+
+                    # check that ballot_length is shorter than total number of cands
+                    # ballot = ballot[: self.ballot_length+1]
+                    ballot_pool.append(ballot)
+
+                # Bloc ballots
+                for _ in range(num_ballots_race - crossover_ballots):
+                    ballot = bloc_cands + opposing_cands
+                    # ballot = ballot[: self.ballot_length+1]
+                    ballot_pool.append(ballot)
+
+        pp = self.ballot_pool_to_profile(
+            ballot_pool=ballot_pool, candidates=self.candidates
+        )
+        return pp
+
+
+class CambridgeSampler(BallotGenerator):
+    def __init__(
+        self,
+        path: Path,
+        slate_to_candidate: dict = {},
+        pref_interval_by_slate: dict = {},
+        slate_voter_prop: dict = {},
+        slate_crossover_rate: dict = {},
+        **data,
+    ):
+        # Call the parent class's __init__ method to handle common parameters
+        super().__init__(**data)
+
+        # Assign additional parameters specific to
+        if not self.parameterized:
+            self.slate_to_candidate = slate_to_candidate
+            self.pref_interval_by_slate = pref_interval_by_slate
+            self.slate_voter_prop = slate_voter_prop
+            self.slate_crossover_rate = slate_crossover_rate
+            self.path = path
+
+    def generate_profile(self) -> PreferenceProfile:
+
+        # Load the ballot_type frequencies used in the Cambridge model
+        # dict of form
+        # {(cand_race1, cand_race1, ...): num_ballots, }
+        # e.g. {('W', 'C', 'C'): 15, ('C', 'W', 'C', 'W'): 9, ...}
+        with open(self.path, "rb") as pickle_file:
+            ballot_frequencies = pickle.load(pickle_file)
+
+        ballot_pool = []
+
+        blocs = self.slate_to_candidate.keys()
+        for bloc in blocs:
+            opp_bloc = next(iter(set(blocs).difference(set(bloc))))
+
+            # bloc or opp-first probabilities for each ballot variant
+            bloc_first_count = sum(
+                [
+                    freq
+                    for ballot, freq in ballot_frequencies.items()
+                    if ballot[0] == bloc
+                ]
+            )
+            prob_ballot_given_bloc_first = {
+                ballot: freq / bloc_first_count
+                for ballot, freq in ballot_frequencies.items()
+                if ballot[0] == bloc
+            }
+            prob_ballot_given_opp_first = {
+                ballot: freq / bloc_first_count
+                for ballot, freq in ballot_frequencies.items()
+                if ballot[0] == opp_bloc
+            }
+
+            bloc_voters = int(self.slate_voter_prop[bloc] * self.number_of_ballots)
+            pref_interval_dict = self.pref_interval_by_slate[bloc]
+            for _ in range(bloc_voters):
+                first_choice = np.random.choice(
```

Comment:
I believe (at a cursory glance at least) this code could be simplified to instead choose between `prob_ballot_given_block_first` and `prob_ballot_given_opp_first` and then the `bloc_ordering = random.choices` call could just use the result of that weighted choice. (removing duplication in lines 468-480)

taking that even further, it doesn't look like both are used, it will always be one or the other.  you could defer the calculation itself, so something like:
```
# make weighted choice between blocs
# calculate the appropriate `prob_ballot`
# select `bloc_ordering` from `prob_ballot`
```